core/src/apps/bitcoin/sign_tx/layout.py

```diff
@@ -59,6 +59,34 @@ async def confirm_output(ctx: wire.Context, output: TxOutput, coin: CoinInfo) ->
     await require_confirm(ctx, text, ButtonRequestType.ConfirmOutput)
 
 
+async def confirm_replacement(ctx: wire.Context, description: str, txid: bytes) -> None:
+    text = Text(description, ui.ICON_SEND, ui.GREEN)
+    text.normal("Confirm transaction ID:")
+    hex_data = hexlify(txid).decode()
+    if len(hex_data) >= 18 * 4:
+        hex_data = hex_data[: (18 * 4 - 3)] + "..."
+    text.mono(*split_op_return(hex_data))
+    await require_confirm(ctx, text, ButtonRequestType.SignTx)
+
+
+async def confirm_modify_fee(
+    ctx: wire.Context, user_fee_change: int, total_fee_new: int, coin: CoinInfo
+) -> None:
+    text = Text("Fee modification", ui.ICON_SEND, ui.GREEN)
+    if user_fee_change == 0:
+        text.normal("Your fee did not change.")
```

Comment:
I consider the wording of the dialog misleading because `user_fee_change` is not the difference between
  * the mining fee in the original transaction(s) and
  * the mining fee in the replacement transaction

but the difference between
  * the user's contribution to the external outputs and the mining fee in the original transaction(s) and
  * the user's contribution to the external outputs and the mining fee in the replacement transaction.

Reply:
the user's contribution to external couldn't have increased

in the typical case, contribution to external outputs is unchanged, so this is indeed the difference in fee

otherwise, the user's contribution to external outputs might be lower, and the difference goes:

* towards the fee - which is a meaningless distinction from the user's POV

* back towards the user, in which case indeed the reported "fee modification" can logically go lower than the original fee.

consider:
*original*:
```
ext_in = 10; ext_out = 20
int_in = 10; int_out = 0; int_contrib = 10
fee = 0
```
*replacement*:
```
ext_in = 15; ext_out = 20
int_in = 10; int_out = 5; int_contrib = 5
fee = 0
```
The calculated `user_fee_change == -5`, but the original fee was zero, so "your fee has decreased" is untrue. "Your contribution was decreased" would be more precise -- but if there are no external signers and the user is sending money back to themselves, "contribution" is a wrong word.

We *could* pass `fee - orig_fee` and a flag indicating whether there are external signers, but I'm honestly not sure if that's worth it.
Maybe instead think of the wording.

Maybe "You are spending <X> more", "You are spending <X> less", "Your spending has not changed"?

also ISTM that "Your fee has increased" and "Your fee did not change" is always correct in referring to a "fee"? And the only problematic case is the decrease.

Reply:
I agree with @matejcik for the most part.

What I don't entirely agree with:
> Maybe "You are spending more", "You are spending less", "Your spending has not changed"?

I wouldn't want to completely replace the "fee" terminology with words like "spending" or "paying" when communicating with the user, because it introduces way more confusion than clarification. As a user I would not necessarily consider the fee to be a part of what I am spending/paying, i.e. I may view the fee as a separate expense. BTW, that's why on the standard confirmation screen we say "including fee ...", because it's not obvious whether it's included or not. So I would hate to mess up the terminology, because of one perhaps confusing but non-dangerous edge case.

To summarize and elaborate a bit on what's happening:
1. If there are no external inputs, then talking about "fee" is always correct.
2. If there are external inputs, then:
    a) If the user is spending more, then talking about "fee" is also correct.
    b) If the user is spending less and
          i. `spending - orig_spending == fee - orig_fee` then talking about "fee" is also correct.
          ii. `-orig_fee <= spending - orig_spending < fee - orig_fee` then talking about "fee" _may_ be incorrect, but not dangerous.
          iii. `-orig_fee > spending - orig_spending` then talking about "fee" _is_ incorrect, but not dangerous.

I can imagine creating a separate dialog `confirm_modify_tx()` for case(s) 2. b) iii. (and possibly ii.) in which I wouldn't mind using the word "spending". So for example "Decrease the amount you are spending by:" and the title of that dialog would be "Tx modification" and it would still show the total fee.

Reply:
> 1. If there are no external inputs, then talking about "fee" is always correct.

in a cancellation transaction with no external inputs, talking about "fee" is strictly incorrect

Reply:
but WRT wording, this folds to the same case as 2b) iii.

Reply:
I certainly agree that if we choose to support cancellation transactions in the future, then those should have a confirmation dialog of their own. In my summary above I am assuming the current implementation, where reducing or removing external outputs is not allowed.

Reply:
As I mentioned at the standup, I would appreciate some feedback on the preferred solution to this. I take it that everybody agrees with my summary. I should add that case 2. b) iii. can be further split into two subcases: `spending >=0` and `spending < 0`. In the latter subcase it doesn't even make sense to talk about "spending less". This can only be described as "receiving". So the dialog should say "Do you agree to receive: ...". This terminology could also be applied to 2. b) iii. and possibly even ii. However, I think we might want to refrain from showing information about the amount being received on Trezor, because that might even be dangerous. I imagine an attacker could use this to trick an inexperienced user into thinking that it is a confirmation of having received a payment. So some options that come to mind are:
1. Show only the TXID dialog(s), but don't show the final fee dialog at all if `spending < orig_spending`, i.e. don't ask user to confirm any fee decreases or receivings. Presumably the user would have no reason to refuse.
2. Show only the TXID dialog(s), but don't show the final fee dialog at all if `spending - orig_spending` and `spending - orig_spending < fee - orig_fee`, i.e. don't ask user to confirm any fee decreases or receivings when new external inputs are involved (2. b) ii. and 2. b) iii.). Basically same as the first option, but limits the rule only to ambiguous situations. 
3. `user_fee_change = max(spending - orig_spending, -orig_fee, - orig_spending)`. The advantage here is that the user would see when the other party in PayJoin offers to pay part of the fees. However, seems like a bad idea when coins with negative fees come into play.

Reply:
I understand why you don't like the word "paying", nevertheless I consider "You are spending more" still better than "Increase your fee by".

Maybe product should decide that. @tsusanka

Reply:
> `spending < 0`

in what sort of circumstances would the user be _signing_ a transaction that is essentially receiving? some sort of mixer?

I mildly dislike not showing a confirmation of what's going on in case the tx confirmations show "fee modification", but I'm not sure if that is happening in the cases under discussion. And oherwise i'm honestly getting somewhat lost.

Going back to my breakdown, i'm offering this: "if there are external outputs, and `spending < orig_spending`, change "your fee has decreased" to "your contribution has decreased" "
But I don't have any strong opinions either way and leaving this up for someone else to decide.

Reply:
> I consider "You are spending more" still better than "Increase your fee by".

On the contrary, "Increase your fee by ..." seems quite obviously better than "Increase the amount you are spending by ...", because the former is correct and provides more accurate and more comprehensive information, as I already explained above (see ambiguity about fee inclusion). The problematic situations appear in the opposite scenario, when the user is getting more in the replacement transaction than in the original transaction and there are new external inputs. In that case "Decrease the amount you are spending by ..." might be better in some instances than "Decrease your fee by ...", but in the end both of these statements are incorrect if the user receives more in the replacement, than what they were spending in the original (see below).

> > `spending < 0`
> 
> in what sort of circumstances would the user be _signing_ a transaction that is essentially receiving? some sort of mixer?

The most typical use-case would be a 3-party PayJoin. Say Alice creates a transaction where she is sending 0.03 BTC to Bob and paying a 0.001 BTC fee. Carol comes along and wants to pay Alice 1 BTC. So Alice gives Carol the transaction and asks her to amend it and sign it. Then Alice signs the amended transaction. In this case it doesn't make sense to talk about the fee being decreased, it doesn't even make sense to talk about the spending amount being decreased.

> I mildly dislike not showing a confirmation of what's going on in case the tx confirmations show "fee modification", but I'm not sure if that is happening in the cases under discussion.

That could happen in the first option I proposed. In the second option that I proposed (_Show only the TXID dialog(s), but don't show the final fee dialog at all if `spending < orig_spending` and `spending - orig_spending < fee - orig_fee`_) the TXID confirmations would show "PayJoin".

> Going back to my breakdown, i'm offering this: "if there are external outputs, and `spending < orig_spending`, change "your fee has decreased" to "your contribution has decreased"

Although I agree that it would be better than talking about "fee", the main problem that I have with this is that in the 3-party PayJoin example above this would show "Decrease your contribution by: 1 BTC", which doesn't make much sense, since Alice put in only 0.031 to begin with. The more sensible thing to show would be "Do you want to receive: 1 BTC", which has problems of it's own, as I already explained (see tricking the user above).

So to summarize, depending on the numbers, it's possible to split this into 3 scenarios: "fee" / "contribution" / "receiving". However, showing the "receiving" confirmation could do way more harm than good, and since there is a thin line between the "contribution" and "receiving" scenario, in my eyes it sort of pulls down the "contribution" scenario as well. So at the moment I am inclining towards one of the following:
1. Not showing the last two scenarios at all, just the TXID confirmation, to avoid doing more damage than good.
2. Using the word "contribution" in the last two scenarios, as @matejcik proposed, in which case we would have to accept that the message doesn't always make perfect sense, e.g. some 3-party CoinJoins.

Reply:
I implemented a solution which simply skips confirmation of the fee in PayJoin if the user is not increasing their contribution. This is a slight variation on what I proposed above and it's easy see that it does not suffer from the concern that @matejcik had:
> I mildly dislike not showing a confirmation of what's going on in case the tx confirmations show "fee modification", but I'm not sure if that is happening in the cases under discussion.

The rules can be summarized once more as:
1. If we are doing an ordinary fee modification or a transaction meld, then ask the user to confirm the fee always.
2. Otherwise, i.e. if we are doing PayJoin, then ask the user to confirm fee modification only when the transaction is increasing their contribution to the fee.
    The rationale behind this is that in the other PayJoin scenarios where confirmation is skipped:
    - There is no risk of coin loss.
    - The scenarios can be open to multiple interpretations and attempting to explain them may cause unnecessary confusion. See above.

Commit: https://github.com/trezor/trezor-firmware/pull/1292/commits/461e44db4b97be369d702fe0ee80580b042c15a0